### PR TITLE
AudioSystem

### PR DESCRIPTION
**Pull request by PiotrMoscicki**

### CODE REVIEW

File: PolyEngine/Engine/Src/SoundWorldComponent.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		/// @see OpenALRenderingDevice
+		SoundWorldComponent(World* world);
+
+		/// Releases stored context.
+		~SoundWorldComponent();
+
+		const ALCcontext* GetContext() const { return Context; }
+
+	private:
+		ALCcontext* Context;
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/SoundWorldComponent.hpp]: Is it really necessary to have one context per world? For graphics it isn't. Please provide some explanation.

Comment (PiotrMoscicki) [inline PolyEngine/Engine/Src/SoundWorldComponent.hpp]: Context in OpenAL is more like World than graphics context; it stores listener and sources. Unless we want to calculate everything and use OpenAL only for playing calculated audio i suggest using one context per world.

Comment (MuniuDev) [inline PolyEngine/Engine/Src/SoundWorldComponent.hpp]: Ok, let's keep it like that then.

File: PolyEngine/Engine/Src/SoundResource.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
+#include "EnginePCH.hpp"
+
+#include <al.h>
+#include <ogg/ogg.h>
+#include <vorbis\vorbisfile.h>
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/SoundResource.cpp]: Please use "/"

File: PolyEngine/Engine/Src/SoundResource.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+		int eos = 0;
+		int i = 0;
+		char* buffer;
+		size_t bytesRead = 0;
+
+		buffer = ogg_sync_buffer(&syncState, 4096);
+		memcpy(buffer, reinterpret_cast<void*>(reinterpret_cast<size_t>(data->GetBegin()) + bytesRead), 4096);
+		bytesRead += 4096;
+		if (bytesRead > data->GetSize()) break;
+		else ogg_sync_wrote(&syncState, 4096);
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/SoundResource.cpp]: codding style.

File: PolyEngine/Core/Src/BinaryBuffer.hpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	public:
+		BinaryBuffer(size_t);
+		~BinaryBuffer();
+
+		void* GetBegin() const { return Begin; }
+		size_t GetSize() const { return Size; }
+		void AddData(void*, size_t);
+
+	private:
+		void* Begin;
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/BinaryBuffer.hpp]: Rename to data. Make it char*.

File: PolyEngine/Core/Src/BinaryBuffer.hpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+namespace Poly
+{
+
+	class CORE_DLLEXPORT BinaryBuffer : public BaseObject<>
+	{
+	public:
+		BinaryBuffer(size_t);
+		~BinaryBuffer();
+
+		void* GetBegin() const { return Begin; }
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/BinaryBuffer.hpp]: change to `char* GetData...`

File: PolyEngine/Core/Src/BinaryBuffer.hpp [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+		size_t GetSize() const { return Size; }
+		void AddData(void*, size_t);
```
Comment (MuniuDev) [inline PolyEngine/Core/Src/BinaryBuffer.hpp]: This shouldn't be necessary. Remove this method. If you need extending buffer functionality use `Dynarray<char>`

File: PolyEngine/Engine/Src/SoundResource.cpp [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+	while (true)
+	{
+		int eos = 0;
+		int i = 0;
+		char* buffer;
+		size_t bytesRead = 0;
+
+		buffer = ogg_sync_buffer(&syncState, 4096);
+		memcpy(buffer, reinterpret_cast<void*>(reinterpret_cast<size_t>(data->GetBegin()) + bytesRead), 4096);
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/SoundResource.cpp]: If previous comments are fixed, then remove these ugly casts.

File: PolyEngine/Engine/Src/SoundResource.cpp [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+using namespace Poly;
+
+SoundResource::SoundResource(const String& path)
+{
+	alGenBuffers(1, &BufferID);
+
+	// Declarations and loading file to buffer.
+
+	BinaryBuffer* data = LoadBinaryFile(path);
+	BinaryBuffer* dupa = new BinaryBuffer(0);
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/SoundResource.cpp]: If you need resizable buffer use `Dynarray<>`

File: PolyEngine/Engine/Src/SoundResource.cpp [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+#include <al.h>
+#include <ogg/ogg.h>
+#include <vorbis\vorbisfile.h>
+#include <fstream>
+
+#include "SoundResource.hpp"
+
+using namespace Poly;
+
+SoundResource::SoundResource(const String& path)
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/SoundResource.cpp]: This whole method seems to be overcomplicated. See this: https://www.gamedev.net/articles/programming/general-and-gameplay-programming/introduction-to-ogg-vorbis-r2031

**Pull request merged**